File: src/py_pubsub/py_pubsub/topics_practice_b.py
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import Twist
from turtlesim.msg import Pose

class TopicsPracticeB(Node):

    def __init__(self):
        super().__init__('topics_practice_b')
        self.publisher_ = self.create_publisher(Twist, 'turtle1/cmd_vel', 10)
        self.pose_subscriber_ = self.create_subscription(Pose, 'turtle1/pose', self.pose_callback, 10)
        self.timer_ = self.create_timer(1.0, self.timer_callback)
        self.countdown = 5
        self.in_spiral_mode = True
        self.linear_velocity = 0.9
        self.angular_velocity = 1.1
        self.current_x = 0.0
        self.current_y = 0.0

    def pose_callback(self, msg):
        self.current_x = msg.x
        self.current_y = msg.y

    def timer_callback(self):
        if self.countdown > 0:
            self.get_logger().info(f'Application starts in: {self.countdown}')
            self.countdown -= 1
        else:
            if self.in_spiral_mode:
                self.draw_spiral()
            else:
                self.go_straight_or_avoid_walls()

    def draw_spiral(self):
        msg = Twist()
        self.linear_velocity += 0.08
        msg.linear.x = self.linear_velocity
        msg.angular.z = self.angular_velocity
        self.publisher_.publish(msg)
        self.get_logger().info('Drawing spiral')
        if self.reached_boundaries():
            self.in_spiral_mode = False

    def go_straight_or_avoid_walls(self):
        if self.near_wall():
            msg = Twist()
            msg.linear.x = 0.0
            msg.angular.z = 1.0
            self.publisher_.publish(msg)
            self.get_logger().info('Avoiding walls')
        else:
            msg = Twist()
            msg.linear.x = self.linear_velocity
            msg.angular.z = 0.0
            self.publisher_.publish(msg)
            self.get_logger().info('Going straight')

    def near_wall(self):
        x_limit = 1.0
        y_limit = 1.0
        if self.current_x < x_limit or self.current_x > 10.5 - x_limit or self.current_y < y_limit or self.current_y > 10.5 - y_limit:
            return True
        return False

    def reached_boundaries(self):
        x_limit = 9.0
        y_limit = 9.0
        if self.current_x > x_limit or self.current_x < 1.0 or self.current_y > y_limit or self.current_y < 1.0:
            return True
        return False

def main(args=None):
    rclpy.init(args=args)
    # Run topics_practice_b node
    topics_practice_b = TopicsPracticeB()
    rclpy.spin(topics_practice_b)
    topics_practice_b.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/cpp_pubsub/src/topics_practice_b.cpp
#include <chrono>
#include <memory>
#include <string>
#include <cmath>

#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/twist.hpp"
#include "turtlesim/msg/pose.hpp"

using namespace std::chrono_literals;

class TopicsPracticeB : public rclcpp::Node
{
public:
    TopicsPracticeB()
    : Node("topics_practice_b"), countdown_(5), in_spiral_mode_(true), linear_velocity_(0.9), angular_velocity_(1.1), current_x_(0.0), current_y_(0.0)
    {
        publisher_ = this->create_publisher<geometry_msgs::msg::Twist>("turtle1/cmd_vel", 10);
        pose_subscriber_ = this->create_subscription<turtlesim::msg::Pose>(
            "turtle1/pose", 10, std::bind(&TopicsPracticeB::pose_callback, this, std::placeholders::_1));
        timer_ = this->create_wall_timer(
            1s, std::bind(&TopicsPracticeB::timer_callback, this));
    }

private:
    void pose_callback(const turtlesim::msg::Pose::SharedPtr msg)
    {
        current_x_ = msg->x;
        current_y_ = msg->y;
    }

    void timer_callback()
    {
        if (countdown_ > 0)
        {
            RCLCPP_INFO(this->get_logger(), "Application starts in: %d", countdown_);
            countdown_--;
        }
        else
        {
            if (in_spiral_mode_)
            {
                draw_spiral();
            }
            else
            {
                go_straight_or_avoid_walls();
            }
        }
    }

    void draw_spiral()
    {
        auto msg = geometry_msgs::msg::Twist();
        linear_velocity_ += 0.08;
        msg.linear.x = linear_velocity_;
        msg.angular.z = angular_velocity_;
        publisher_->publish(msg);
        RCLCPP_INFO(this->get_logger(), "Drawing spiral");
        if (reached_boundaries())
        {
            in_spiral_mode_ = false;
        }
    }

    void go_straight_or_avoid_walls()
    {
        if (near_wall())
        {
            auto msg = geometry_msgs::msg::Twist();
            msg.linear.x = 0.0;
            msg.angular.z = 1.0;
            publisher_->publish(msg);
            RCLCPP_INFO(this->get_logger(), "Avoiding walls");
        }
        else
        {
            auto msg = geometry_msgs::msg::Twist();
            msg.linear.x = linear_velocity_;
            msg.angular.z = 0.0;
            publisher_->publish(msg);
            RCLCPP_INFO(this->get_logger(), "Going straight");
        }
    }

    bool near_wall()
    {
        const double x_limit = 1.0;
        const double y_limit = 1.0;
        return (current_x_ < x_limit || current_x_ > 10.5 - x_limit || current_y_ < y_limit || current_y_ > 10.5 - y_limit);
    }

    bool reached_boundaries()
    {
        const double x_limit = 9.0;
        const double y_limit = 9.0;
        return (current_x_ > x_limit || current_x_ < 1.0 || current_y_ > y_limit || current_y_ < 1.0);
    }

    rclcpp::TimerBase::SharedPtr timer_;
    rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr publisher_;
    rclcpp::Subscription<turtlesim::msg::Pose>::SharedPtr pose_subscriber_;
    int countdown_;
    bool in_spiral_mode_;
    double linear_velocity_;
    double angular_velocity_;
    double current_x_;
    double current_y_;
};

int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv);
    auto topics_practice_b = std::make_shared<TopicsPracticeB>();
    rclcpp::executors::MultiThreadedExecutor executor;
    executor.add_node(topics_practice_b);
    executor.spin();
    rclcpp::shutdown();
    return 0;
}